enhancements/ocp-coreos-layering/ocp-coreos-layering.md

```diff
@@ -103,23 +127,201 @@ A future enhancement will help automate the above.  For example, we may support
+
+###### Create Final Pool Images
+When the Customer Base Image has been successfully tagged into cluster, a final image will be created for each pool.
+- The Final Pool Image is cluster & pool specific
+- It includes templates and other cluster pool specific content which gets layered on top of the Customer Base Image.
+- By default the Customer Base Image will be applied to all pools unless the user specifies a target pool. There will also be an imagestream per pool that can be used for overrides as well.
+- The user can watch the pool to keep track of build progress for the pool image to troubleshoot in case build fails.
+- Final pool image generated from this build will be also available to customer to pull and test. This exact image will be used by MCO to update nodes in the desired pools.
+- There are some [implementation details](#how-the-final-image-will-be-built) that will need to be solved.
+
+**Note**: This text is proposing a new CRD and some custom flow.  Another alternative is having the final pool image be pushed to an imagestream, and simply using pool pause to control final rollout.
+
+###### Rollout of Final Image to Node
+- Once the built final pool image has been tested by the user, they can initiate a rollout to a specific pool:
+   ```
+   $ oc rollout custombuild mcp/worker
+   This is a disruptive action and will replace the current rhcos image with a custom build. Are you sure?
+  ```
+- Rollout follows the existing MCO rollout paradigm - i.e. the images roll out to max unavailable nodes in a pool. This will cause reboots.
+- After rollout has finished, there is an easy way to see which builds are present in the pools:
+  ```
+  $ oc get custombuilds -n openshift-machine-config-operator
+  BUILD         POOL         BASEIMAGE     STATUS        LOCATION
+  cp-layers    master        “quay.io…”    Pending          TBD
+  w-layers     worker        “quay.io…”    Done         image-registry...
+  ```
+
+###### Reverting to Standard OS
+- At any given time, the user should be able to revert to the stock rhcos image for that cluster which comes in via release image:
+  ```
+  $ oc delete custombuild mcp/worker
+  This is a disruptive action and will replace custom build XXXX.XX with rhelcoreos 89.12. Are you sure?
+  ```
+- This action will remove all custom layers and cause nodes to use the image coming in through osimageurl with no modifications.
+- Non-negotiable: roll-backs to stock must be tested and supported
+
+###### Providing Customer Base Image during cluster upgrade
+- OS can be updated with Customer Base Image during a cluster upgrade. For this, the user will need to create the Customer Base Image using the incoming rhcos version and supply to upgrade so that default osimageurl is not used:
+  ```
+  $ oc add upgrade —layeredsource -f
```

Comment:
Could this be a ConfigMap that has a reference to the version the cluster is updating to and an os image url that is built for that version?